cachi2/core/package_managers/pip.py

```diff
@@ -1601,52 +1630,97 @@ def _validate_provided_hashes(requirements: list[PipRequirement], require_hashes
-    :raises PackageRejected: if sdists for the package is not found or yanked
-    """
-    timeout = get_config().requests_timeout
-    package = requirement.package
+def _process_package_distributions(
+    requirement: PipRequirement,
+    pip_deps_dir: RootedPath,
+    user_checksums: dict[tuple[str, str], set[ChecksumInfo]] = dict(),
+    allow_binary: bool = False,
+) -> tuple[DistributionPackageInfo, list[DistributionPackageInfo]]:
+    name = requirement.package
     version = requirement.version_specs[0][1]
 
-    # See https://www.python.org/dev/peps/pep-0503/
-    package_url = f"{pypi_url.rstrip('/')}/simple/{canonicalize_name(package)}/"
+    normalized_name = canonicalize_name(name)
+    normalized_version = canonicalize_version(version)
+
+    client = pypi_simple.PyPISimple()
     try:
-        pypi_resp = pkg_requests_session.get(package_url, auth=pypi_auth, timeout=timeout)
-        pypi_resp.raise_for_status()
-    except requests.RequestException as e:
+        timeout = get_config().requests_timeout
+        project_page = client.get_project_page(name, timeout)
+        packages = project_page.packages
+    except (requests.RequestException, pypi_simple.NoSuchProjectError) as e:
         raise FetchError(f"PyPI query failed: {e}")
 
-    html = bs4.BeautifulSoup(pypi_resp.text, "html.parser")
-    # Find all anchors anywhere in the doc, the PEP does not specify where they should be
-    links = html.find_all("a")
+    def _to_dpi(dp: pypi_simple.DistributionPackage, with_digests: bool) -> DistributionPackageInfo:
+        dpi = DistributionPackageInfo(
+            name, version, pip_deps_dir.join_within_root(dp.filename).path, dp.url, dp.is_yanked
+        )
+        if with_digests:
+            for algorithm, digest in dp.digests.items():
+                dpi.checksums.add(ChecksumInfo(algorithm, digest))
+        return dpi
+
+    wheels = []
+    if allow_binary:
+        filtered = [
+            p
+            for p in packages
+            if p.version
+            and canonicalize_version(p.version) == normalized_version
+            and p.package_type == "wheel"
+        ]
+
+        user_req_checksums = user_checksums.get((normalized_name, normalized_version), set())
+        if not user_req_checksums:
+            wheels = [_to_dpi(w, with_digests=True) for w in filtered]
+        else:
+            for w in filtered:
+                dpi = _to_dpi(w, with_digests=False)
+
+                for algorithm, digest in w.digests.items():
+                    checksum = ChecksumInfo(algorithm, digest)
+                    if checksum in user_req_checksums:
+                        dpi.checksums.add(checksum)
+
+                if dpi.checksums:
+                    wheels.append(dpi)
+                else:
+                    log.info(f"Filtering out {w.filename}: checksum does not match")
+
+            if not wheels:
+                raise PackageRejected("", solution="")
```

Comment:
Let's not fail here, as long as there's at least one archive (including the sdist) the build has a chance to work.

Let's just log a warning if there's at least 1 wheel in `filtered` but 0 wheels in `wheels`